File: cases/android-whitelabel/app/src/main/java/com/posthog/whitelabel/CheckoutService.kt
// Fails identically in every brand.
package com.posthog.whitelabel

class CheckoutException(message: String) : Exception(message)

object CheckoutService {
    fun buyPremium() {
        validateCart()
    }

    private fun validateCart() {
        chargeCard()
    }

    private fun chargeCard() {
        throw CheckoutException("Payment declined (code 402)")
    }
}

File: cases/ios-whitelabel/Whitelabel/CheckoutService.swift
//
//  CheckoutService.swift
//
//  Fails identically in every brand. @inline(never) keeps the call chain
//  visible in Release stack traces.
//

import Foundation

enum CheckoutError: Error {
    case paymentDeclined(code: Int)
}

enum CheckoutService {
    @inline(never)
    static func buyPremium() throws {
        try validateCart()
    }

    @inline(never)
    private static func validateCart() throws {
        try chargeCard()
    }

    @inline(never)
    private static func chargeCard() throws {
        throw CheckoutError.paymentDeclined(code: 402)
    }
}
